### PR TITLE
storage: add OriginTimestamp to MVCCValueHeader

### PR DESCRIPTION
**Pull request by msbutler**
This patch adds a new OriginTimestamp field to the MVCCValueHeader which will identify the timestamp the kv was written on the original source cluster during Logical Data Replication. By default (in non LDR contexts), this field is nil'd out.

In future patches, we will actually bind the origin timestamp here during LDR ingestion and use it to conduct large scale fingerprinting and LWW conflict resolution.

Epic: none

Release note: none

### CODE REVIEW

File: pkg/storage/enginepb/mvcc3.proto [hunk 1]
```diff
@@ -213,8 +213,14 @@ message MVCCValueHeader {
   // when LDR is not being run, 1 identifies a remote write, and 2+ are reserved
   // to identify remote clusters.
   uint32 origin_id = 5  [(gogoproto.customname) = "OriginID"];
+  
+  // OriginTimestamp identifies the timestamp this kv was written on the
+  // original source cluster during Logical Data Replication. A default empty
```
Comment (stevendanna) [inline pkg/storage/enginepb/mvcc3.proto:218]: Maybe 
```suggestion
  // OriginTimestamp is the MVCC timestamp of the originating cluster's KV 
  // for values written by Logical Data Replication. A default empty
```

File: pkg/storage/enginepb/mvcc3.proto [hunk 2]
```diff
@@ -213,8 +213,14 @@ message MVCCValueHeader {
...
+  // timestamp implies that this kv did not originate from a Logical Data
+  // Replication stream.
+  util.hlc.Timestamp origin_timestamp = 6  [(gogoproto.casttype) = "github.com/cockroachdb/cockroach/pkg/util/hlc.Timestamp"];
 
-  // NextID = 6.
+   // NextID = 7.
```
Comment (stevendanna) [inline pkg/storage/enginepb/mvcc3.proto:223]: [nit] extra space

**Review by stevendanna — COMMENTED** [review]
LGTM but we should definitely get multiple reviews on these PRs that touch storage.

One thing we should think about is whether these new fields should be included in Export requests.

Another thing we might want to consider is that between this and OriginID, we've nearly doubled the size of the MVCCValueHeader struct in go.

Last release's addition of ImportEpoch was probably free in terms of struct space because of the previous padding.  But the OriginID and OriginTimestamp will definitely take space. I suppose we could reclaim some of this space if we went with a struct that we then stored a pointer to.  That would incur an extra allocation on unmarshaling of the items.

None of this matters for _storage_ space so perhaps it isn't worth worrying about.

**Review by RaduBerinde — COMMENTED** [review]
To add to Steven's point, the new timestamp is a pointer anwyay so we will incur that extra allocation as things stand. Could we move both `OriginID` and `OriginTimestamp` to a separate struct and hold a pointer to that? (the OriginTimestamp would be non-nullable within that struct)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127386#-:-O25L8UUBvbOt9dPwPYD:bg3ec9s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt, @jbowens, and @msbutler)

**Comment by msbutler** [comment]
> Could we move both OriginID and OriginTimestamp to a separate struct and hold a pointer to that? (the OriginTimestamp would be non-nullable within that struct)

@RaduBerinde steven and I discussed offline that that's what we plan to do. however, i'm planning to open a preliminary PR that moves the current OriginID field to a separate struct and backport this patch to 24.2. This backport should not have any side effects, except that LDR can't be run from the current 24.2 beta to the next 24.2 beta. We'd like to do this backport so 24.2 LDR is not completely incompatible with 24.3 LDR. does that all make sense?

**Comment by RaduBerinde** [comment]
Absolutely, sounds great!

**Comment by msbutler** [comment]
opened https://github.com/cockroachdb/cockroach/pull/127465

**Pull request closed without merging**